pkgs/top-level/all-packages.nix

```diff
@@ -606,11 +606,20 @@ in
   makeWrapper = makeSetupHook { deps = [ dieHook ]; substitutions = { shell = targetPackages.runtimeShell; }; }
                               ../build-support/setup-hooks/make-wrapper.sh;
 
-  makeBinaryWrapper = makeSetupHook {
-    deps = [ dieHook ];
-  } (runCommand "make-binary-wrapper.sh" {} ''
-      substitute ${../build-support/setup-hooks/make-binary-wrapper.sh} $out --replace " cc " " ${gcc}/bin/cc "
-  '');
+  makeBinaryWrapper = let
+    f = { cc, sanitizers }: let
+      san = lib.concatMapStringsSep " " (s: "-fsanitize=${s}") sanitizers;
+      script = runCommand "make-binary-wrapper.sh" {} ''
+        substitute ${../build-support/setup-hooks/make-binary-wrapper.sh} $out \
+          --replace " @CC@ " " ${cc}/bin/cc ${san} "
+      '';
+    in
+      makeSetupHook { deps = [ dieHook ]; } script;
+  in
+    lib.makeOverridable f {
+      cc = gcc;
+      sanitizers = [ "undefined" "address" ];
```

Comment:
don't think these should be enabled by default for all packages, only in its own tests. but having the option is very useful.

Reply:
What are the downsides to having them enabled by default? Will it significantly impact performance/binary size?

Reply:
Please leave them enabled by default. I prefer having sanitizers explode loudly instead of having the program doing wild things in case we have undefined behavior triggered in a raw c program. Users who really want to squeeze the last bit performance can still consciously disable this.

Reply:
@tfc we agree with you, but sanitizers are a huge overhead for short binaries. we've run a test of a binary that's just ten putenvs, and sanitizers slowed it down 25-fold (0.4ms per run without sanitizers, 10ms with sanitizers enabled). compared to bash wrappers it's still a 3x slowdown (3.5ms for bash). we would argue that that's just *too much* overhead for something that can be exhaustively tested, like the current wrappers can be. if there was an option to include custom code we'd *totally* want those sanitizers to be enabled at all times when custom code is involved though.

Reply:
FWIW, on my M1 Pro, benchmarked with `hyperfine`:

* No sanitizers: 0.4 ms ± 0.3 ms
* With `-fsanitize=address` alone: 50.6 ms ± 0.8 ms
* With `-fsanitize=undefined` alone: 2.6 ms ± 0.4 ms
* With both: 51.7 ms ± 1.3 ms

So perhaps we can keep `-fsanitize=undefined` by default?

Reply:
if it's cheaper than bash it's a win :+1: would still prefer them off by default though. the code is small and readily audited/tested, and we currently trust bash to not attend any shenanigans on odd inputs too.

Reply:
Yes, given that these programs do not process input of any kind, I also don't see why the sanitizers should be on by default. Having them in the tests seem sufficient. (This is also in line with how other programs use sanitizers.)

Reply:
This fails because `stdenv.cc.cc` is unwrapped and has no `/bin/cc`